samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java

```diff
@@ -443,6 +443,18 @@ public void run() {
             long startTime = clock.nanoTime();
             task.window(coordinator);
             containerMetrics.windowNs().update(clock.nanoTime() - startTime);
+
+            // We have scenario when the task window method takes longer time than window trigger time, there is a
+            // potential for process method starvation. Ideally, we need to merge the thread pool for commit, window with
+            // process to guarantee fairness. In the interim, we want to warn users about potential starvation.
```

Comment:
This comment is not very accurate. Commit,window and process are running in the same thread pool if user sets the job.thread.pool.size. The reason why window is always running is that we give priority to fire window when we schedule the next operation for a task ( see AsyncTaskState.nextOp()). This is to guarantee the window will fire on time.